test/functional/wallet_transactiontime_rescan.py

```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/env python3
+# Copyright (c) 2018-2019 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test transaction time during old block rescanning
+"""
+
+import time
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+    assert_equal
+)
+
+
+class TransactionTimeRescanTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = False
+        self.num_nodes = 3
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def run_test(self):
+        self.log.info('Prepare nodes and wallet')
+
+        minernode = self.nodes[0]  # node used to mine BTC and create transactions
+        usernode = self.nodes[1]  # user node with correct time
+        restorenode = self.nodes[2]  # node used to restore user wallet and check time determination in ComputeSmartTime (wallet.cpp)
+
+        # time constant
+        cur_time = int(time.time())
+        ten_days = 10 * 24 * 60 * 60
+
+        # synchronize nodes and time
+        self.sync_all()
+        minernode.setmocktime(cur_time)
+        usernode.setmocktime(cur_time)
+        restorenode.setmocktime(cur_time)
+
+        # prepare miner wallet
+        minernode.createwallet(wallet_name='default')
+        miner_wallet = minernode.get_wallet_rpc('default')
+        m1 = miner_wallet.getnewaddress()
+
+        # prepare the user wallet with 3 watch only addresses
+        wo1 = usernode.getnewaddress()
+        wo2 = usernode.getnewaddress()
+        wo3 = usernode.getnewaddress()
+
+        usernode.createwallet(wallet_name='wo', disable_private_keys=True)
+        wo_wallet = usernode.get_wallet_rpc('wo')
+
+        wo_wallet.importaddress(wo1)
+        wo_wallet.importaddress(wo2)
+        wo_wallet.importaddress(wo3)
+
+        self.log.info('Start transactions')
+
+        # check blockcount
+        assert_equal(minernode.getblockcount(), 200)
+
+        # generate some btc to create transactions and check blockcount
+        minernode.generatetoaddress(300, m1)
```

Comment:
I think 101 blocks would suffice here (the test functions the same but would run faster), unless I'm missing something.
 
<details><summary>working example</summary><p>

```diff
         # generate some btc to create transactions and check blockcount
-        minernode.generatetoaddress(300, m1)
-        assert_equal(minernode.getblockcount(), 500)
+        initial_mine = 101
+        minernode.generatetoaddress(initial_mine, m1)
+        assert_equal(minernode.getblockcount(), initial_mine + 200)
 
         # synchronize nodes and time
         self.sync_all()
@@ -75,7 +76,7 @@ class TransactionTimeRescanTest(BitcoinTestFramework):
 
         # generate blocks and check blockcount
         minernode.generatetoaddress(100, m1)
-        assert_equal(minernode.getblockcount(), 600)
+        assert_equal(minernode.getblockcount(), initial_mine + 300)
 
         # synchronize nodes and time
         self.sync_all()
@@ -88,7 +89,7 @@ class TransactionTimeRescanTest(BitcoinTestFramework):
 
         # generate blocks and check blockcount
         minernode.generatetoaddress(100, m1)
-        assert_equal(minernode.getblockcount(), 700)
+        assert_equal(minernode.getblockcount(), initial_mine + 400)
 
         # synchronize nodes and time
         self.sync_all()
@@ -101,7 +102,7 @@ class TransactionTimeRescanTest(BitcoinTestFramework):
 
         # generate more blocks and check blockcount
         minernode.generatetoaddress(100, m1)
-        assert_equal(minernode.getblockcount(), 800)
+        assert_equal(minernode.getblockcount(), initial_mine + 500)
 
         # check user final balance and transaction count
         self.log.info('Check user final balance')
@@ -136,7 +137,7 @@ class TransactionTimeRescanTest(BitcoinTestFramework):
 
         # proceed to rescan, first with an incomplete one, then with a full rescan
         self.log.info('Rescan last history part')
-        restorewo_wallet.rescanblockchain(650)
+        restorewo_wallet.rescanblockchain(initial_mine + 350)
         self.log.info('Rescan all history')
         restorewo_wallet.rescanblockchain()
```
</p></details>

Reply:
done.